hackathon_site/dashboard/frontend/src/slices/hardware/hardwareSlice.ts

```diff
@@ -0,0 +1,44 @@
+import { createSlice, createEntityAdapter } from "@reduxjs/toolkit";
+
+export interface Hardware {
+    id: number;
+    name: string;
+    model_number: string;
+    manufacturer: string;
+    datasheet: string;
+    quantity_available: number;
+    notes: string;
+    max_per_team: number;
+    picture: string;
+    categories: number[];
+    quantity_remaining: string;
+}
+
+export interface HardwareState {
+    isLoading: boolean;
+    error: string | null;
+    next: string | null;
+}
+
+export const hardwareReducerName = "hardware";
+const hardwareAdapter = createEntityAdapter();
+export const initialState: HardwareState = {
+    isLoading: false,
+    error: null,
+    next: null,
+};
+
+// Slices
+const hardwareSlice = createSlice({
+    name: hardwareReducerName,
+    initialState: hardwareAdapter.getInitialState(initialState),
+    reducers: {},
+});
+
+export const { actions, reducer } = hardwareSlice;
+export default reducer;
+
+// Selectors
+export const hardwareSliceSelector = (state) => state[hardwareReducerName];
```

Comment:
`state` here needs to be typed with the type of our root state (I'm getting an error otherwise).

You'll have to convert `slices/store.js` to `slices/store.ts`. You should only get one error about `history` being untyped in `rootReducer`, just add `History` to the import at the top from `"history"` and do

```typescript
const rootReducer = (history: History) => ...
```

After that, you can export the type of the root state at the bottom:

```typescript
export type RootState = ReturnType<typeof store.getState>;
```

Then in this file:

```typescript
import { RootState } from "slices/store";
...

export const hardwareSliceSelector = (state: RootState) => state[hardwareReducerName];
```